### DISCUSSION TITLE
How to use WSL as backend for Android development?

### DISCUSSION THREAD

**Discussion by diskria** [microsoft/WSL#10347] [Q&A] [answered]
Is it possible to use WSL as backend for development? I imagine it like you are running Android Studio on Windows side, using a fully functional file system without X-server. but all tools, java and gradle are on the WSL side and compile the project. 

The fact is that I don't like setting up the X server, the inability to transfer the file using drag&drop to the IDE window, and much more. But at the same time, I can't switch completely to Windows, since the build speed there is 2 times slower. But I don't want to switch to Linux OS either.

**Comment by diskria** [comment, marked as answer]
I got it

---

**Hybrid Development Setup with Android Studio and WSL**

In an effort to optimize the development environment for Android applications on a Windows machine, I implemented a hybrid setup that utilizes both Windows and Windows Subsystem for Linux (WSL). This configuration allows me to write code in Windows while compiling and building in WSL, leveraging the strengths of both environments for increased efficiency.

**Syncing Code with Rsync**

The core of this setup is the use of `rsync` to synchronize the project directory in Windows with WSL. This synchronization happens every time the project needs to be built, ensuring that all recent code changes in the Windows environment are mirrored in WSL. I customized `rsync` to exclude directories that either contain a large number of files or only need updates when the application version changes. This selective syncing prevents unnecessary transfer of build caches, local configurations, and the `.git` directory, thereby speeding up the synchronization process.

**Optimizing Rsync Conditions**

The synchronization script checks for version changes to determine the intensity of the syncing needed—if only minor changes are detected, a quick sync is triggered, otherwise a full sync is performed. This conditional logic optimizes the sync process based on the nature of the updates, saving time and computing resources.

**Building with Gradle in WSL**

Once the codebase is updated in WSL, I execute the build using Gradle, passing the same command-line arguments (`-P`) that Android Studio normally uses. This ensures that the build process in WSL is consistent with what would occur in a purely Windows-based environment.


**Post-Build copy APK**

After the build process is completed in WSL, instead of synchronizing the entire project directory back to Windows, I simply copy the final APK file. This streamlined approach focuses exclusively on transferring the APK from WSL to the Windows environment. By doing so, I ensure that the newly generated APK, which is the direct output of the compilation, is immediately available in Windows for testing or deployment purposes. This method avoids the overhead of syncing other build artifacts or modified source files, making the workflow more efficient and tailored specifically to the needs of deploying or testing the application.

**Integration into Android Studio**

To streamline the process, I integrated this setup into Android Studio as an External Tool. This external tool replaces the default Gradle Aware Make in configuration allowing me to build projects directly from the IDE with WSL as the backend. This setup maintains the native IDE experience, including error reporting and debugging tools.

**Handling File Paths for IDE Compatibility**

Additionally, I implemented a script to parse the build output and convert WSL file paths back to their Windows equivalents. This ensures that any clickable file links in the Android Studio logs point correctly to the source files on Windows, facilitating a seamless development experience.

**Conclusion**

This hybrid approach not only improves build times significantly but also retains the convenience and features of Android Studio on Windows. By offloading the build process to WSL, I effectively leverage the strengths of both operating systems, leading to a more efficient development cycle.

---

This setup could be particularly useful for developers looking to enhance their productivity without sacrificing the familiar tools and environment provided by Android Studio.

**Reply by david-stroup** [reply]
Thank you! Could you share more about your setup? Maybe compelte `rsync` program which you used and External tool ssetup for Android Studio
